macrosynergy/panel/basket_performance_class.py

```diff
@@ -0,0 +1,390 @@
+
+import numpy as np
+import pandas as pd
+import random
+from typing import List
+from macrosynergy.panel.historic_vol import expo_weights, expo_std, flat_std
+from macrosynergy.management.shape_dfs import reduce_df_by_ticker
+from macrosynergy.panel.converge_row import ConvergeRow
+
+class Basket(object):
+
+    def __init__(self, df: pd.DataFrame, contracts: List[str], ret: str = "XR_NSA",
+                 cry: str = None, start: str = None, end: str = None,
+                 blacklist: dict = None, wgt: str = None):
+
+        """
+        Class' Constructor.
+
+        :param <pd.Dataframe> df: standardized DataFrame with following columns: 'cid',
+            'xcats', 'real_date' and 'value'.
+        :param <List[str]> contracts: base tickers (combinations of cross-sections and
+            base categories) denoting contracts that go into the basket.
+        :param <str> ret: return category postfix; default is "XR_NSA".
+        :param <str> cry: carry category postfix; default is None.
+        :param <str> start: earliest date in ISO 8601 format. Default is None.
+        :param <str> end: latest date in ISO 8601 format. Default is None.
+        :param <dict> blacklist: cross-sections with date ranges that should be excluded
+            from the dataframe. If one cross-section has several blacklist periods append
+            numbers to the cross-section code.
+        :param <str> wgt: postfix used to identify exogenous weight category. Analogously
+            to carry and return postfixes this should be added to base tickers to
+            identify the values that denote contract weights. Only applicable for the
+            weight methods 'values' or 'inv_values'.
```

Comment:
The class should calculate weight dataframes for all methods. Even equal weights change overtime based on the availability of returns/carries (big issue!) and blacklisting.